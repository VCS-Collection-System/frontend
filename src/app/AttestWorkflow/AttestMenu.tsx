import React, {useEffect, useState} from 'react';
import {
  Button,
  Card,
  CardBody,
  CardFooter,
  Stack,
  Text,
} from "@patternfly/react-core";
import { CheckCircleIcon, CogIcon, FileUploadIcon } from '@patternfly/react-icons';
import { useHistory } from 'react-router-dom';
import CardWithTitle from '@app/Components/CardWithTitle';
import Feature, { showFeature } from '@app/Components/Feature';
import { useKeycloak } from "@react-keycloak/web";
import { getEmployeeInfo } from '@app/utils/getEmployeeInfo';

const AttestMenu: React.FunctionComponent = () => {
    const history = useHistory();
    const { keycloak } = useKeycloak();
    const employee = getEmployeeInfo(keycloak);
    const [ isRecovery, setIsRecovery ] = useState(false); 

    useEffect(() => {
      showFeature("recovery", keycloak).then(config => {
        setIsRecovery(config);
      });
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);
    
    return (
      <Card isRounded>
        <CardWithTitle title="Menu" info="Please select from the following:" />
        <CardBody>
          <Stack hasGutter>
            <Button
              icon={<FileUploadIcon />}
              id="submit-vax-card-menu-button"
              isSmall
              onClick={() => history.push("/attest/vax")}
              variant="tertiary"
            >
              <Text>Submit Vaccination Record</Text>
              <Feature name="proofOptional" renderOff={() => (
                <Text>Supporting Document Upload Required</Text>
              )}>
                &nbsp;
              </Feature>
            </Button>
            <Feature name="covidTest" renderOff={() =>
              <Feature name="recovery"> 
                <Button
                  icon={<FileUploadIcon />}
                  id="submit-vax-test-menu-button"
                  isSmall
                  onClick={() => history.push("/attest/test")}
                  variant="tertiary"
                >
                  <Text>Submit Recovery Certificate</Text>
                  <Text>Supporting Document Upload Required</Text>
                </Button>
              </Feature>
            }> 
              <Button
                icon={<FileUploadIcon />}
                id="submit-vax-test-menu-button"
                isSmall
                onClick={() => history.push("/attest/test")}
                variant="tertiary"
              >
                <Text>Submit Test Results{isRecovery ? " or Recovery Certificate" : ""}</Text>
                <Text>Supporting Document Upload Required</Text>
              </Button>
            </Feature>
            {employee.isApprover ? (
              <Feature name="hrDashboard"> 
                <Button
                  icon={<CheckCircleIcon />}
                  id="validate-submissions-button"
                  isSmall
                  onClick={() => history.push("/hrdashboard")}
                  variant="tertiary"
                >
                  <Text>HR Dashboard</Text>
                  <Text>Review or Update Red Hat VCS Records</Text>
                </Button>
              </Feature>
            ) : ""}
            {employee.isSuperApprover ? (
              <Button
                  icon={<CogIcon />}
                  id="settings-button"
                  isSmall
                  onClick={() => history.push("/settings")}
                  variant="tertiary"
                >
                  <Text>Application Settings</Text>
                  <Text>Update Red Hat VCS Country Settings</Text>
                </Button>
            ) : ""}
          </Stack>
        </CardBody>
        <CardFooter>
          <Button 
            variant="plain" 
            id="attest-menu-back-button"
            onClick={() => history.push("/profile")}
          >
            Back
          </Button>
        </CardFooter>
      </Card>
    );
};
export { AttestMenu }
