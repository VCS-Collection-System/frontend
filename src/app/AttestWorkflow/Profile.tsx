import React from "react";
import {
  Alert,
  Button,
  Card,
  CardBody,
  Checkbox,
  Form,
  FormGroup,
  TextInput,
} from "@patternfly/react-core";
import { CheckCircleIcon } from "@patternfly/react-icons";
import CardWithTitle from "@app/Components/CardWithTitle";
import { useForm } from "react-hook-form";
import { useHistory } from "react-router-dom";
import { useKeycloak } from "@react-keycloak/web";
import { getEmployeeInfo } from "@app/utils/getEmployeeInfo";
import { BLOCKED_COUNTRIES } from "@app/utils/constants";

const Profile: React.FunctionComponent = () => {
  const history = useHistory();
  const { handleSubmit } = useForm();
  const { keycloak } = useKeycloak();
  const employee = getEmployeeInfo(keycloak);
  const employeeFirstName = employee.firstName;
  const employeeLastName = employee.lastName;
  const employeeEmail = employee.email;
  const employeeCountry = employee.country;

  const onSubmit = () => {
    history.push("/attestmenu");
  };

  return (
    <Card isRounded>
      {typeof(employeeCountry) === "undefined" || BLOCKED_COUNTRIES.includes(employeeCountry) ? (
        <Alert variant="danger" title="This application is not currently available in your region." style={{marginBottom: "var(--pf-global--gutter--md)"}}>
          Please visit our <a href="/contact">Help</a> page for more information.
        </Alert>
      ) : ""}
      <CardWithTitle
        title="Profile"
        info="The information below is retrieved from your Red Hat user account."
      />
      <CardBody>
        <Form onSubmit={handleSubmit(onSubmit)}>
          <FormGroup
            id="employee-full-name"
            label="Full Name"
            fieldId="profile-form-employee-full-name"
          >
            <TextInput
              aria-label="employee-full-name"
              isDisabled
              type="text"
              value={(employeeFirstName || "") + " " + (employeeLastName || "")}
            />
          </FormGroup>
          <FormGroup
            id="employee-email-display"
            label="Email"
            fieldId="profile-form-email"
          >
            <TextInput
              aria-label="employee-email-display"
              isDisabled
              type="email"
              value={employeeEmail || ""}
            />
          </FormGroup>
          {employee.isApprover && 
            (
              <Checkbox
                id="reviewer-checkbox"
                label="Red Hat VCS Reviewer"
                aria-label="disabled checked checkbox"
                defaultChecked
                isDisabled
              />
          )}
          <FormGroup fieldId="accept-test-buttons">
            <Button 
              icon={<CheckCircleIcon />} 
              variant="primary" 
              type="submit"
              isDisabled={typeof(employeeCountry) === "undefined" || BLOCKED_COUNTRIES.includes(employeeCountry)}
              >
              Accept
            </Button>
          </FormGroup>
        </Form>
      </CardBody>
    </Card>
  );
};

export { Profile };
