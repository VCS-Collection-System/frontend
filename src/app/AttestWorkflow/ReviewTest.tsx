import React, { useState } from 'react';
import {
  Button, 
  Card,
  CardFooter,
  CardBody,
  Form,
  FormGroup,
  TextInput
} from '@patternfly/react-core';
import { ExternalLinkAltIcon, OutlinedPaperPlaneIcon } from '@patternfly/react-icons';
import { useHistory } from 'react-router-dom';
import CardWithTitle from '@app/Components/CardWithTitle';
import { convertBase64ToBlob, dateConvert } from "@app/utils/utils";
import authaxios from "@app/utils/axiosInterceptor";
import { ResponseErrorModal } from '@app/Components/FeedbackModals';
import Feature from "@app/Components/Feature";
import { useKeycloak } from "@react-keycloak/web";
import { getEmployeeInfo } from '@app/utils/getEmployeeInfo';
import { SUBMIT_TEST_URL } from '@app/utils/constants';

const ReviewTest: React.FunctionComponent = () => {
  const [isErrorModalOpen, setIsErrorModalOpen] = useState(false); 
  const [errorState, setErrorState] = useState('');
  const [isFormSubmitting, setIsFormSubmitting] = useState(false);
  const history = useHistory();
  const { keycloak } = useKeycloak();
  const employee = getEmployeeInfo(keycloak);
  const employeeFirstName = employee.firstName;
  const employeeLastName = employee.lastName;
  const employeeID = employee.id;
  const employeeEmail = employee.email;
  const employeeCountry = employee.country;
  const employeeName = (employee.firstName || "") + " " + (employee.lastName || ""); 

  const testImage = sessionStorage.getItem("test_image_file") || "";
  const testImageFilename = sessionStorage.getItem("test_image_filename") || "";
  const altFirstName = sessionStorage.getItem("alt_first_name") || "";
  const altLastName = sessionStorage.getItem("alt_last_name") || "";
  const employeeDOB = sessionStorage.getItem("date_of_birth") || "";
  const covidTestResult = sessionStorage.getItem("test_results") || "";
  const covidTestDate = sessionStorage.getItem("test_date") || "";
  const altName = (altFirstName ? altFirstName : employeeFirstName) + " " +  (altLastName ? altLastName : employeeLastName);

  const handleSubmit = ()=> {
    setIsFormSubmitting(true);
    const testCard = convertBase64ToBlob(testImage)
    const formData = new FormData();

    const employeeInfo = {
      id: employeeID,
      firstName: employeeFirstName,
      lastName: employeeLastName,
      altFirstName: altFirstName,
      altLastName: altLastName,
      dateOfBirth: employeeDOB,
      email: employeeEmail,
      agencyCode: employeeCountry,
      agencyName: employeeCountry,
      divisionCode: ""
    };

    const testInfo = {
      covidTestDate: covidTestDate,
      covidTestResult: covidTestResult.toUpperCase()
    }

    const attachmentExtension = testImageFilename.split('.').pop()?.toLowerCase();
    let attachmentType;

    if(attachmentExtension === "pdf") {
      attachmentType = "application/pdf";
    }
    else if(attachmentExtension === "jpg") {
      attachmentType = "image/jpeg";
    }
    else {
      attachmentType = "image/" + attachmentExtension;
    }

    formData.append('employee', new Blob([JSON.stringify(employeeInfo)], {type: "application/json"}) );
    formData.append('documents', new Blob([JSON.stringify(testInfo)], {type: "application/json"}));
    formData.append("attachment", new Blob([testCard], { type: attachmentType }), "blob."+attachmentExtension);

    authaxios.post(SUBMIT_TEST_URL, formData, {
      headers: {
        'Content-Type': 'multipart/form-data;'
      }
    })
    .then(() => {
      setIsFormSubmitting(false);
      history.push('/thankyou');
    })
    .catch(res => {
      if(res.response) {
        setIsFormSubmitting(false);
        setErrorState(res.response.status);
        setIsErrorModalOpen(true);
      }
    })
  }

  return (
    <Card isRounded>
      <ResponseErrorModal 
        isModalOpen={isErrorModalOpen} 
        setIsModalOpen={setIsErrorModalOpen} 
        errorState={errorState}
      />
      <CardWithTitle
        title="Review Submission"
        info="Review your test result submission and validate that test result image matches displayed data. Please press Submit below to proceed."
      />
      <CardBody>
        <Form>
          <FormGroup label="Person ID" fieldId="review-test-form-employee-id">
            <TextInput
              aria-label="review-test-form-employee-id"
              isDisabled
              type="text"
              value={employeeID}
            />
          </FormGroup>
          <FormGroup
            fieldId="review-test-form-employee-full-name"
            label="Full Name"
          >
            <TextInput
              aria-label="review-test-form-employee-full-name"
              isDisabled
              type="text"
              value={employeeName}
            />
          </FormGroup>
          <FormGroup
            fieldId="review-test-form-employee-alt-name"
            label="Full Preferred Name (as displayed on test result)"
          >
            <TextInput
              aria-label="review-test-form-employee-alt-name"
              isDisabled
              type="text"
              value={altName}
            />
          </FormGroup>
          <Feature name="dateOfBirth">
            <FormGroup
              label="Date of Birth"
              fieldId="review-test-date-of-birth"
            >
              <TextInput
                aria-label="review-test-date-of-birth"
                isDisabled
                type="text"
                value={dateConvert(employeeDOB)}
              />
            </FormGroup>
          </Feature>
          <FormGroup label="Test Results" fieldId="review-test-form-results">
            <TextInput
              aria-label="review-test-form-results"
              isDisabled
              type="text"
              value={sessionStorage.getItem("test_results") || ""}
            />
          </FormGroup>
          <FormGroup label="Test Date" fieldId="review-test-form-date">
            <TextInput
              aria-label="review-test-form-date"
              isDisabled
              type="text"
              value={dateConvert(sessionStorage.getItem("test_date") || "")}
            />
          </FormGroup>
          <FormGroup label="Image preview" fieldId="review-vax-image-preview">
            {testImage && testImageFilename.split('.').pop()?.toLowerCase() == "pdf" &&
              (typeof testImage !== "undefined" ? (
                <Button
                  icon={<ExternalLinkAltIcon />}
                  variant="link"
                  component="a"
                  href={URL.createObjectURL(convertBase64ToBlob(testImage))}
                  target="_blank"
                >
                  Test Result PDF
                </Button>
              ) : null)}
            {testImage && testImageFilename.split('.').pop()?.toLowerCase() != "pdf" &&
              (typeof testImage !== "undefined" ? (
                <div>
                  <img
                    src={URL.createObjectURL(convertBase64ToBlob(testImage))}
                    alt="test-results-image"
                  />
                </div>
              ) : null)}
          </FormGroup>
        </Form>
      </CardBody>
      <CardFooter>
        <Button
          icon={<OutlinedPaperPlaneIcon />}
          isDisabled={isFormSubmitting}
          isLoading={isFormSubmitting}
          onClick={handleSubmit}
          variant="primary"
        >
          Submit
        </Button>
        <Button onClick={() => history.push("/attest/test")} variant="plain">
          Back
        </Button>
      </CardFooter>
    </Card>
  );
};

export { ReviewTest };
