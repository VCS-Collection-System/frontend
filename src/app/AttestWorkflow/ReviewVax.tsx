import React, { useState } from "react";
import {
  Card,
  CardFooter,
  CardBody,
  Button,
  Form,
  FormGroup,
  Grid,
  TextInput
} from "@patternfly/react-core";
import { ExternalLinkAltIcon, OutlinedPaperPlaneIcon } from '@patternfly/react-icons';
import { useHistory } from "react-router-dom";
import { convertBase64ToBlob, dateConvert } from "@app/utils/utils";
import authaxios from "@app/utils/axiosInterceptor";
import { ResponseErrorModal } from '@app/Components/FeedbackModals';
import { useKeycloak } from "@react-keycloak/web";
import CardWithTitle from "@app/Components/CardWithTitle";
import ReviewVaxDose from "@app/Components/ReviewVaxDose";
import Feature from "@app/Components/Feature";
import { getEmployeeInfo } from "@app/utils/getEmployeeInfo";
import { SUBMIT_VAX_URL } from "@app/utils/constants"; 

const ReviewVax: React.FunctionComponent = () => {
  const [isErrorModalOpen, setIsErrorModalOpen] = useState(false);
  const [errorState, setErrorState] = useState("");
  const [isFormSubmitting, setIsFormSubmitting] = useState(false);
  const altFirstName = sessionStorage.getItem("alt_first_name") || "";
  const altLastName = sessionStorage.getItem("alt_last_name") || "";
  const employeeDOB = sessionStorage.getItem("date_of_birth") || "";
  const proofType = sessionStorage.getItem("proof_type") || "";
  const vaxImage = sessionStorage.getItem("vax_image") || "";
  const vaxFilename = sessionStorage.getItem("vax_image_filename") || "";

  const doses: {number: number, date: string, lotNumber: string, vaccine: string}[] = [];

  for(let i = 1; sessionStorage.getItem("vax_date_" + i) != null && sessionStorage.getItem("vax_date_" + i) != ""; i++) {
    doses.push({
      number: i,
      date: sessionStorage.getItem("vax_date_" + i) || "",
      lotNumber: sessionStorage.getItem("lot_number_" + i) || "",
      vaccine: sessionStorage.getItem("vaccine_" + i) || ""
    });
  }

  const history = useHistory();
  const { keycloak } = useKeycloak();
  const employee = getEmployeeInfo(keycloak);
  const employeeFirstName = employee.firstName;
  const employeeLastName = employee.lastName;
  const employeeID = employee.id;
  const employeeEmail = employee.email;
  const employeeCountry = employee.country;

  const altName = (altFirstName ? altFirstName : employeeFirstName) + " " +  (altLastName ? altLastName : employeeLastName);

  const formSubmitSuccess = () => {
    setIsFormSubmitting(false);
    history.push("/thankyou");
  }
  const handleSubmit = () => {
    setIsFormSubmitting(true);

    const employeeInfo = {
      id: employeeID,
      firstName: employeeFirstName,
      lastName: employeeLastName,
      altFirstName: altFirstName,
      altLastName: altLastName,
      dateOfBirth: employeeDOB,
      email: employeeEmail,
      agencyCode: employeeCountry,
      agencyName: employeeCountry,
      divisionCode: ""
    };

    const vaxInfos = doses.map((dose) => {
      return {
        vaccineBrand: dose.vaccine.toUpperCase(),
        vaccineAdministrationDate: dose.date,
        vaccineShotNumber: dose.number,
        lotNumber: dose.lotNumber,
        proofType: proofType
      };
    });
    
    const formData = new FormData();

    formData.append("employee", new Blob([JSON.stringify(employeeInfo)], { type: "application/json" }));
    formData.append("documents", new Blob([JSON.stringify(vaxInfos)], { type: "application/json" }));
    if(vaxFilename !== "") {
      const vaxCard = convertBase64ToBlob(vaxImage);
      const attachmentExtension = vaxFilename.split('.').pop()?.toLowerCase();
      let attachmentType;

      if(attachmentExtension === "pdf") {
        attachmentType = "application/pdf";
      }
      else if(attachmentExtension === "jpg") {
        attachmentType = "image/jpeg";
      }
      else {
        attachmentType = "image/" + attachmentExtension;
      }

      formData.append("attachment", new Blob([vaxCard], { type: attachmentType }), "blob."+attachmentExtension);
    } else {
      formData.append('attachment', new Blob([], { type: "image/png" }), "blob.png");
    }

    const headers = {
      "Content-Type": "multipart/form-data;",
    };

    authaxios.post(SUBMIT_VAX_URL, formData, {
      headers: headers
    }).then(() => {
      formSubmitSuccess();
    }).catch((res) => {
      if (res.response) {
        setIsFormSubmitting(false);
        setErrorState(res.response.status);
        setIsErrorModalOpen(true);
      }
    });
  }

  const recordOptions = [
    { value: "CDC", label: "CDC Vaccination Card" },
    { value: "DIVOC", label: "Digital Certification of Vaccination" },
    { value: "GREEN_PASS", label: "Green Pass" },
    { value: "OTHER", label: "Other" }
  ];

  const proofPreview = recordOptions.find(option => option.value === proofType) || {};

  return (
    <Card isRounded>
      <ResponseErrorModal 
        isModalOpen={isErrorModalOpen} 
        setIsModalOpen={setIsErrorModalOpen} 
        errorState={errorState}
      />
      <CardWithTitle
        title="Review Vaccination Submission"
        info="Please review your vaccination record submission and validate that card image matches displayed data or it will be rejected. Please press Submit below to proceed."
      />
      <CardBody>
        <Form>
          <FormGroup
            fieldId="review-vax-form-employee-full-name"
            label="Full Name"
          >
            <TextInput
              aria-label="review-vax-form-employee-full-name"
              isDisabled
              type="text"
              value={employeeFirstName + " " + employeeLastName}
            />
          </FormGroup>
          <FormGroup
            fieldId="review-vax-form-employee-alt-name"
            label="Full Preferred Name (as displayed on vaccination record)"
          >
            <TextInput
              aria-label="review-vax-form-employee-alt-name"
              isDisabled
              type="text"
              value={altName}
            />
          </FormGroup>
          <Feature name="dateOfBirth">
            <FormGroup
              label="Date of Birth"
              fieldId="review-vax-form-date-of-birth"
            >
              <TextInput
                aria-label="review-vax-form-date-of-birth"
                isDisabled
                type="text"
                value={dateConvert(employeeDOB)}
              />
            </FormGroup>
          </Feature>
          <Grid hasGutter>
            {doses.map((dose, i) => (
              <ReviewVaxDose doseNumber={dose.number} key={i} date={dateConvert(dose.date)} lotNumber={dose.lotNumber} vaccine={dose.vaccine} />
            ))}
          </Grid>
          <Feature name="proof">
            <FormGroup label="Vaccination Record Type" fieldId="review-vax-form-vaccine-proof">
              <TextInput
                aria-label="review-vax-form-vaccine-proof"
                isDisabled
                type="text"
                value={proofPreview['label']}
              />
            </FormGroup>
            <FormGroup label="Attachment preview" fieldId="review-vax-image-preview">
              {vaxImage && vaxFilename.split('.').pop()?.toLowerCase() == "pdf" && (typeof vaxImage !== "undefined" ? (
                  <Button
                    icon={<ExternalLinkAltIcon />}
                    variant="link"
                    component="a"
                    href={URL.createObjectURL(convertBase64ToBlob(vaxImage))}
                    target="_blank"
                  >
                    Vaccination Card PDF
                  </Button>
              ) : null)}
              {vaxImage && vaxFilename.split('.').pop()?.toLowerCase() != "pdf" &&
                (typeof vaxImage !== "undefined" ? (
                  <div>
                    <img
                      src={URL.createObjectURL(convertBase64ToBlob(vaxImage))}
                      alt="vaccination-card"
                    />
                  </div>
                ) : null)}
              {!vaxImage && (
                <p>No image attached.</p>
              )}
            </FormGroup>
          </Feature>
        </Form>
      </CardBody>
      <CardFooter>
        <Button
          icon={<OutlinedPaperPlaneIcon />}
          isDisabled={isFormSubmitting}
          isLoading={isFormSubmitting}
          variant="primary"
          onClick={handleSubmit}
        >
          Submit
        </Button>
        <Button variant="plain" onClick={() => history.push("/attest/vax")}>
          Back
        </Button>
      </CardFooter>
    </Card>
  );
};

export { ReviewVax };
