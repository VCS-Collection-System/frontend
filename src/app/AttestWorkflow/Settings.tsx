import React, { useEffect, useState } from 'react';
import { useForm, Controller } from "react-hook-form";
import { useKeycloak } from '@react-keycloak/web';
import { useHistory } from "react-router-dom";
import { 
  Button,
  Card,
  CardBody,
  CardTitle,
  Form,
  FormGroup,
  FormSelect,
  FormSelectOption,
  Modal,
  ModalVariant,
  Select,
  SelectOption,
  SelectVariant } from '@patternfly/react-core';
import { CheckCircleIcon } from '@patternfly/react-icons';
import CardWithTitle from '@app/Components/CardWithTitle';
import { showFeature } from '@app/Components/Feature';
import SettingsSwitch from '@app/Components/SettingsSwitch';
import { getEmployeeInfo } from "@app/utils/getEmployeeInfo";
import authaxios from "@app/utils/axiosInterceptor";
import { SAVE_CONFIG_URL } from '@app/utils/constants';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const countries = require("../Assets/countries.json");

const options = [<SelectOption value="default" key={0}>Global Standard</SelectOption>].concat(
  countries.map((country, i) => (
    <SelectOption value={country["alpha-3"]} key={i+1}>
      {country["name"]}
    </SelectOption>
  ))
)

const Settings: React.FunctionComponent = () => {
  const history = useHistory();
  const { keycloak } = useKeycloak();
  const form = useForm();
  const [ proofEnabled, setProofEnabled ] = useState(false);
  const [ hrEnabled, setHREnabled ] = useState(false);
  const [ isSelectOpen, setIsSelectOpen ] = useState(false);
  const [ isModalOpen, setIsModalOpen ] = useState(false);
  
  const userCountry = getEmployeeInfo(keycloak).country;
  const [ country, setCountry ] = useState(userCountry ? userCountry : "default");


  const isEnabled = async (feature) => {
    return await showFeature(feature, keycloak, form.getValues("country"));
  }

  const resetValues = () => {
    let countryConfig = sessionStorage.getItem("globalconfig-" + form.getValues("country"));
    if(countryConfig) {
      countryConfig = JSON.parse(countryConfig);
    }

    for(const field in form.getValues()) {
      if(field !== "country" && field.search("-gs") < 0) {
        isEnabled(field).then(value => {
          form.setValue(field, value);
          const switchElement = document.getElementById(field + "-switch") as HTMLFormElement;
          if(switchElement && countryConfig && countryConfig[field] != null) {
            form.setValue(field + "-gs", false);
            // manually enable because setValue doesn't trigger onChange
            switchElement.disabled = false;
          } else if(switchElement) {
            form.setValue(field + "-gs", true);
            // manually disaable because setValue doesn't trigger onChange
            switchElement.disabled = true;
          }
        });
      }
    }
  }

  useEffect(() => {
    resetValues();

    showFeature("proof", keycloak, country).then(config => {
      setProofEnabled(config);
    });

    showFeature("hrDashboard", keycloak, country).then(config => {
      setHREnabled(config);
    });
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [country]);

  const onSubmit = (data) => {
    const formData = new FormData();
    const configData = {};
    
    for(const field in data) {
      if(field === "country") {
        // all db gets/persists must use lowercase country
        configData["agencyName"] = data["country"].toLowerCase();
        continue;
      }

      // skip global standard checkbox fields
      if(field.indexOf("-gs") > 0) {
        continue;
      }

      if(data["country"] !== "default" && form.getValues(field + "-gs")) {
        // set to null if using global standard
        configData[field] = null;
      } else {
        configData[field] = data[field];
      }
    }

    formData.append("configuration", new Blob([JSON.stringify(configData)], { type: "application/json" }));

    authaxios.post(SAVE_CONFIG_URL, formData).then(() => {
      sessionStorage.setItem("globalconfig-" + country, JSON.stringify(configData));
      setIsModalOpen(true);
    });
  }

  return (
    <>
      <Card isRounded>
        <CardWithTitle
          title="Application Settings"
          info="Update global default or country-specific settings."
        />
        <CardBody>
          <Form onSubmit={form.handleSubmit(onSubmit)} onKeyDown={e => {if(e.code === "Enter") e.preventDefault()}}>
            <Controller
              name="country"
              control={form.control}
              defaultValue={userCountry ? userCountry : "default"}
              render={({ field }) => (
                <FormGroup
                  fieldId="country-select"
                  label="Country"
                >
                  <Select
                    variant={SelectVariant.typeahead}
                    onToggle={isOpen => setIsSelectOpen(isOpen)}
                    onSelect={(event, value) => {
                      field.onChange(value);
                      setCountry(value);
                      setIsSelectOpen(false);
                    }}
                    onFilter={(event, value) => {
                      if (!value || value === "") {
                        return options;
                      }

                      const input = new RegExp(value, 'i');
                      return options.filter((option) => {
                        if(option.props) {
                          return input.test(option.props.value) || input.test(option.props.children);
                        }

                        return false;
                      });
                    }}
                    selections={field.value}
                    isOpen={isSelectOpen}
                    typeAheadAriaLabel="country-select"
                    style={{maxHeight: "400px", overflowY: "scroll"}}
                  >
                    {options}
                  </Select>
                </FormGroup>
              )}
            />
            <Card>
              <CardTitle>Submissions</CardTitle>
              <CardBody>
                <SettingsSwitch
                  name="consentCheckbox"
                  label="Require acknowledgment of employee consent"
                  form={form}
                />
                <SettingsSwitch
                  name="consentCheckboxAlternate"
                  label="Require acknowledgement of employee consent (alternate message)"
                  form={form}
                />
                <SettingsSwitch
                  name="dateOfBirth"
                  label="Enable collection of employee date of birth"
                  form={form}
                />
                <SettingsSwitch
                  name="covidTest"
                  label="Enable submission of COVID test results"
                  form={form}
                />
                <SettingsSwitch
                  name="recovery"
                  label="Enable submission of Certificates of Recovery"
                  form={form}
                />
              </CardBody>
            </Card>
            <Card>
              <CardTitle>Documentation</CardTitle>
              <CardBody>
                <SettingsSwitch
                  name="proof"
                  label="Enable uploading of vaccination proof"
                  onChange={(value) => {
                    setProofEnabled(value);
                  }}
                  form={form}
                />
                <SettingsSwitch
                  name="proofOptional"
                  label="Require vaccination proof"
                  invert
                  hide={!proofEnabled}
                  form={form}
                />
                <SettingsSwitch
                  name="allowPdf"
                  label="Allow PDF files as proof"
                  hide={!proofEnabled}
                  form={form}
                />
                <CardTitle style={{paddingLeft: 0, paddingTop: "3rem", paddingBottom: 0, opacity: !proofEnabled ? ".25" : "1"}}>Proof Types</CardTitle>
                <SettingsSwitch
                  name="cdc"
                  label="Allow CDC Vaccination Record"
                  hide={!proofEnabled}
                  form={form}
                />
                <SettingsSwitch
                  name="greenPass"
                  label="Allow Green Pass"
                  hide={!proofEnabled}
                  form={form}
                />
                <SettingsSwitch
                  name="divoc"
                  label="Allow DIVOC"
                  hide={!proofEnabled}
                  form={form}
                />
                <SettingsSwitch
                  name="eugc"
                  label="Allow EUGC"
                  hide={!proofEnabled}
                  form={form}
                />
              </CardBody>
            </Card>
            <Card>
              <CardTitle>HR Dashboard</CardTitle>
              <CardBody>
                <SettingsSwitch
                  name="hrDashboard"
                  label="Enable HR Dashboard"
                  onChange={(value) => {
                    setHREnabled(value);
                  }}
                  form={form}
                />
                <SettingsSwitch
                  name="hrSearch"
                  label="Enable HR Search"
                  hide={!hrEnabled}
                  form={form}
                />
              </CardBody>
            </Card>
            <FormGroup fieldId="accept buttons">
              <Button 
                icon={<CheckCircleIcon />} 
                type="submit" 
                variant="primary"
              >
                Save
              </Button>
              <Button 
                onClick={() => history.push("/attestmenu")} 
                variant="plain"
              >
                Back
              </Button>
            </FormGroup>
          </Form>
        </CardBody>
      </Card>
      <Modal
        title="Configuration Saved"
        data-testid="saved"
        variant={ModalVariant.small}
        isOpen={isModalOpen}
        onClose={() => {
          setIsModalOpen(false);
        }}
        actions={[
          <Button key="confirm" variant="primary" onClick={() => {setIsModalOpen(false)}}>
            OK
          </Button>
        ]}
      >
        Settings for {form.getValues("country")} saved successfully.
      </Modal>
    </>
  )
};

export { Settings };

