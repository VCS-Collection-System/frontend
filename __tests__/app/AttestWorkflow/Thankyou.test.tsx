import React from 'react';
import { Thankyou } from '../../../src/app/AttestWorkflow/Thankyou';
import { fireEvent, render, screen } from '@testing-library/react'
import mockKeycloak from "../../../__mocks__/keycloakMock"

jest.mock("@react-keycloak/web", () => ({
  useKeycloak: jest.fn(() => ({ keycloak: mockKeycloak })),
}))

describe('<Thankyou />', ()=> {
  it('should be defined', () => {
    expect(Thankyou).toBeDefined()
  })

  it("should match existing snapshot", () => {
    expect(render(<Thankyou />).asFragment()).toMatchSnapshot()
  })

  it("should logout and clear session", () => {
    const mockLogout = jest.fn(); 
    sessionStorage.setItem("mockKey", "mockValue")
    mockKeycloak.logout = mockLogout 
    render(<Thankyou />)

    const logoutButton = screen.getByRole('button', {name: /Finish and Log Out/i})
    fireEvent.click(logoutButton)

    expect(sessionStorage.getItem("mockKey")).toEqual(null);
    expect(mockLogout).toBeCalledTimes(1); 
  })
})