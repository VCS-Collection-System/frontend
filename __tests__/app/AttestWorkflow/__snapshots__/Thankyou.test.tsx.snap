// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Thankyou /> should match existing snapshot 1`] = `
<DocumentFragment>
  <article
    class="pf-c-card pf-m-rounded"
    data-ouia-component-id="OUIA-Generated-Card-1"
    data-ouia-component-type="PF4/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-c-card__title"
    >
      <p
        class=""
        data-pf-content="true"
      >
        Thank You!
      </p>
      <div
        class="pf-c-content"
      >
        <small
          class=""
          data-pf-content="true"
        >
          Thank you for your submission. Your record was successfully submitted.
        </small>
      </div>
    </div>
    <div
      class="pf-c-card__footer"
    >
      <button
        aria-disabled="false"
        class="pf-c-button pf-m-danger"
        data-ouia-component-id="OUIA-Generated-Button-danger-1"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe="true"
        type="button"
      >
        <span
          class="pf-c-button__icon pf-m-start"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 512 512"
            width="1em"
          >
            <path
              d="M464 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h416c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-83.6 290.5c4.8 4.8 4.8 12.6 0 17.4l-40.5 40.5c-4.8 4.8-12.6 4.8-17.4 0L256 313.3l-66.5 67.1c-4.8 4.8-12.6 4.8-17.4 0l-40.5-40.5c-4.8-4.8-4.8-12.6 0-17.4l67.1-66.5-67.1-66.5c-4.8-4.8-4.8-12.6 0-17.4l40.5-40.5c4.8-4.8 12.6-4.8 17.4 0l66.5 67.1 66.5-67.1c4.8-4.8 12.6-4.8 17.4 0l40.5 40.5c4.8 4.8 4.8 12.6 0 17.4L313.3 256l67.1 66.5z"
            />
          </svg>
        </span>
        Finish and Log Out
      </button>
    </div>
  </article>
</DocumentFragment>
`;
